core/arch/arm/pta/sta_benchmark.c

```diff
@@ -0,0 +1,160 @@
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <compiler.h>
+#include <kernel/pseudo_ta.h>
+#include <malloc.h>
+#include <mm/tee_pager.h>
+#include <mm/tee_mm.h>
+#include <mm/core_memprot.h>
+#include <string.h>
+#include <string_ext.h>
+#include <stdio.h>
+#include <trace.h>
+
+#define TA_NAME		"benchmark.ta"
+
+#define BENCHMARK_UUID \
+		{ 0x0b9a63b0, 0xb4c6, 0x4c85, \
+		{ 0xa2, 0x84, 0xa2, 0x28, 0xef, 0x54, 0x7b, 0x4e } }
+
+#define BENCHMARK_CMD(id)			(0xFA190000 + (id & 0xFFFF))
+#define BENCHMARK_CMD_REGISTER_MEMREF	BENCHMARK_CMD(1)
+#define BENCHMARK_CMD_GET_MEMREF	BENCHMARK_CMD(2)
+#define BENCHMARK_CMD_UNREGISTER	BENCHMARK_CMD(3)
+
+void *bench_ts_buf;
+void *bench_stat_buf;
+
+static TEE_Result register_benchmark_memref(uint32_t type,
+				TEE_Param p[TEE_NUM_PARAMS])
+{
+	if ((TEE_PARAM_TYPE_GET(type, 0) != TEE_PARAM_TYPE_MEMREF_INOUT) ||
+		(TEE_PARAM_TYPE_GET(type, 1) != TEE_PARAM_TYPE_MEMREF_INOUT) ||
+		(TEE_PARAM_TYPE_GET(type, 2) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 3) != TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	DMSG("Registering timestamp buffer, addr = %p, paddr = %p\n",
+			p[0].memref.buffer,
+			(void *)virt_to_phys((void *)p[0].memref.buffer));
+	bench_ts_buf = (void *)p[0].memref.buffer;
+
+	DMSG("Registering stat buffer, addr = %p, paddr = %p\n",
+			p[1].memref.buffer,
+			(void *)virt_to_phys((void *)p[1].memref.buffer));
+	bench_stat_buf = (void *)p[1].memref.buffer;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_benchmark_memref(uint32_t type,
+				TEE_Param p[TEE_NUM_PARAMS])
+{
+	if ((TEE_PARAM_TYPE_GET(type, 0) != TEE_PARAM_TYPE_VALUE_OUTPUT) ||
+		(TEE_PARAM_TYPE_GET(type, 1) != TEE_PARAM_TYPE_VALUE_OUTPUT) ||
+		(TEE_PARAM_TYPE_GET(type, 2) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 3) != TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	DMSG("Sending back timestamp buffer paddr, paddr = %p\n",
+		(void *)virt_to_phys(bench_ts_buf));
+	p[0].value.a = virt_to_phys(bench_ts_buf);
+	p[1].value.a = virt_to_phys(bench_stat_buf);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result unregister_benchmark(uint32_t type,
+				TEE_Param p[TEE_NUM_PARAMS] __unused)
+{
+	if ((TEE_PARAM_TYPE_GET(type, 0) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 1) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 2) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 3) != TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	DMSG("Unregistering benchmark, timestamp buffer paddr = %p\n",
+		(void *)virt_to_phys(bench_ts_buf));
+	bench_ts_buf = 0;
+	bench_stat_buf = 0;
```

Comment:
minor: `NULL` rather than `0`. debug trace displays only the timestamp buffer, not the statistics one.